Couple interpolated string fixes

333fred opened the pull request:
Fix a couple of small issues with interpolated string constants, include an insertion break (the definite assignment issue).

Fixes https://github.com/dotnet/roslyn/issues/49720.

333fred commented:
@dotnet/roslyn-compiler for review please.

AlekseyTs commented:
>                             if (value.ConstantValue == null ||

Do we need to execute code starting from here and to the end of the containing block if ```isResultConstant``` is false at this point.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:90 in 8f123bc. [](commit_id = 8f123bc6b96001c1e3fb433d7fe062abdf23e2fb, deletion_comment = False)

333fred commented:
>                             if (value.ConstantValue == null ||

No, I'll add `!isResultConstant ||` to the start of this condition.

---
In reply to: [736887571](https://github.com/dotnet/roslyn/pull/49707#issuecomment-736887571) [](ancestors = 736887571)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:90 in 8f123bc. [](commit_id = 8f123bc6b96001c1e3fb433d7fe062abdf23e2fb, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:
```diff
@@ -724,6 +724,15 @@ internal static bool WriteConsideredUse(TypeSymbol type, BoundExpression value)
                 return value.ConstantValue != ConstantValue.Null;
             }
 
+            if (type?.SpecialType == SpecialType.System_String && value.Kind == BoundKind.InterpolatedString)
```
>if (type?.SpecialType == SpecialType.System_String && value.Kind == BoundKind.InterpolatedString) [](start = 12, length = 97)

Since this is a patch for a very specific scenario when an interpolated string has a constant value, I think the logic should identify the situation precisely and shouldn't return early, but rather adjust the ```if (value.ConstantValue != null) return false;``` to not return for an interpolated string. It also feels like it should be sufficient to simply check for ```value.Kind == BoundKind.InterpolatedString``` and not bother checking the type.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:
```diff
@@ -3756,6 +3756,17 @@ void M2(string S1 = $""Testing"", object O = null, Namae N = null)
+            CreateCompilation(@"
+class C
+{
+    const string s = $"""";
```
>const string s = $""""; [](start = 4, length = 23)

It feels like we should run and verify the value at runtime.

AlekseyTs replied:
Or verify the constant value after importing from metadata.

---
In reply to: [533796984](https://github.com/dotnet/roslyn/pull/49707#discussion_r533796984) [](ancestors = 533796984)

AlekseyTs commented:
Done with review pass (iteration 2)

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -127,6 +121,11 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+            {
+                Debug.Assert(node.Contents.Count == 0);
+                resultConstant = ConstantValue.Create(string.Empty);
+            }
```
Consider checking for `node.Contents.Count == 0` early in the method and avoiding `GetSpecialType()` for `System.Int32` and `System.Object`.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:
```diff
@@ -22,112 +16,122 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
-            var objectType = GetSpecialType(SpecialType.System_Object, diagnostics, node);
-            var intType = GetSpecialType(SpecialType.System_Int32, diagnostics, node);
             ConstantValue? resultConstant = null;
             bool isResultConstant = true;
```
>bool isResultConstant = true; [](start = 12, length = 29)

It looks like `isResultConstant` can be moved inside `else`.

333fred replied:
I'm asserting the result at the bottom, after the else. I'd prefer to keep it where it is to ensure that the assert applies to all paths through the function.

---
In reply to: [533804280](https://github.com/dotnet/roslyn/pull/49707#discussion_r533804280) [](ancestors = 533804280)

cston replied:
It feels like the assert is only needed in the `else` case.

---
In reply to: [533804696](https://github.com/dotnet/roslyn/pull/49707#discussion_r533804696) [](ancestors = 533804696,533804280)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [736892932](https://github.com/dotnet/roslyn/pull/49707#issuecomment-736892932) [](ancestors = 736892932)

AlekseyTs reviewed (approved):
LGTM (iteration 4)

AlekseyTs commented on src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:
```diff
@@ -4842,10 +4842,11 @@ namespace N
 
     class C
     {
+        private const FormattableString {|Rename:Formattable|} = $"""";
```
>private const FormattableString {|Rename:Formattable|} = $""""; [](start = 8, length = 63)

This doesn't look right. 
1) Why it became a field?
2) It feels like this is not going to compile.

333fred replied:
You're right, I didn't review the diff well. Will take a look.

AlekseyTs commented on src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs:
```diff
@@ -7830,10 +7831,11 @@ public void M(string s)
 @"
 public class P
 {
+    private const string {|Rename:V|} = $"""";
```
>private const string {|Rename:V|} = $""""; [](start = 4, length = 42)

It is not clear why this became a field.

333fred replied:
This one is correct. The string has a constant value, and the IDE prefers to extract a constant if it can do so.

AlekseyTs replied:
> This one is correct. The string has a constant value, and the IDE prefers to extract a constant if it can do so.

A constant doesn't have to be a field.

---
In reply to: [533865977](https://github.com/dotnet/roslyn/pull/49707#discussion_r533865977) [](ancestors = 533865977)

AlekseyTs replied:
I would expect it to remain a local.

---
In reply to: [533866290](https://github.com/dotnet/roslyn/pull/49707#discussion_r533866290) [](ancestors = 533866290,533865977)

333fred replied:
> I would expect it to remain a local.

That's a question for the IDE team. They already approved this behavior when other tests were updated with non-empty interpolated strings: [`43b2c9c` (#46828)](https://github.com/dotnet/roslyn/pull/46828/commits/43b2c9c942395b539c02edd72be9aeea154f9d02)

333fred commented:
@dotnet/roslyn-ide for a review of the IDE changes.

CyrusNajmabadi commented on src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService.cs:
```diff
@@ -44,6 +44,9 @@ protected override bool IsInNonFirstQueryClause(ExpressionSyntax expression)
         }
 
+        protected override bool IsInterpolatedStringExpression(ExpressionSyntax expression)
+            => expression.IsKind(SyntaxKind.InterpolatedStringExpression);
```
can be done without an abstrct/override.  just get teh ISyntaxKindsService and check the kind of the node against syntaxKinds.InterpolatedStringExpression

CyrusNajmabadi commented on src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs:
```diff
@@ -97,7 +97,24 @@ public State(TService service, SemanticDocument document)
+                        // Interpolated strings can have constant values, but if it's being converted to a FormattableString
+                        // or IFormattable then we cannot treat it as one
+                        var typeInfo = Document.SemanticModel.GetTypeInfo(Expression, cancellationToken);
+                        IsConstant = typeInfo.ConvertedType?.IsFormattableStringOrIFormattable() != true;
```
this bug probably exists in a *lot* of places.  that said, it feels pretty minor, so i don't ind spot-fixes for it.

333fred replied:
Yep. This particular function is only used in a few places, but I figured that I may as well hit them all at once.

CyrusNajmabadi commented on src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs:
```diff
@@ -97,7 +97,24 @@ public State(TService service, SemanticDocument document)
+                else
+                {
+                    IsConstant = false;
+                }
```
could you extract thi all out to a bool-returning helper method (in this type)

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:
```diff
@@ -231,9 +231,9 @@ public static bool IsAttribute(this ITypeSymbol symbol)
+        public static bool IsFormattableStringOrIFormattable([NotNullWhen(returnValue: true)] this ITypeSymbol? symbol)
         {
-            return symbol?.MetadataName == "FormattableString"
+            return symbol?.MetadataName is "FormattableString" or "IFormattable"
```
nameof?

CyrusNajmabadi commented on src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.State.cs:
```diff
@@ -97,7 +97,24 @@ public State(TService service, SemanticDocument document)
-                IsConstant = Document.SemanticModel.GetConstantValue(Expression, cancellationToken).HasValue;
+                if (Document.SemanticModel.GetConstantValue(Expression, cancellationToken) is { HasValue: true, Value: var value })
+                {
+                    if (_service.IsInterpolatedStringExpression(Expression) && value is string)
```
you have a Documetn here, so you can def access ISyntaxKindsService.  so def remove that absdtract helper.

CyrusNajmabadi reviewed (approved):
LGTM if feedback is addressed.  Thanks!
